src/app/MessageDef.h

```diff
@@ -0,0 +1,1854 @@
+ *      Interaction model profile.
+ *
+ */
+
+#ifndef _CHIP_INTERACTION_MODEL_MESSAGE_DEF_H
+#define _CHIP_INTERACTION_MODEL_MESSAGE_DEF_H
+
+#include <core/CHIPCore.h>
+#include <core/CHIPTLV.h>
+#include <support/CodeUtils.h>
+#include <support/logging/CHIPLogging.h>
+#include <util/basic-types.h>
+
+namespace chip {
+namespace app {
+/**
+ *  @brief
+ *    The CHIP interaction model message types.
+ *
+ *  These values are called out in CHIP Interaction Model: Encoding Specification
+ *
+ */
+enum
+{
+    kMsgType_SubscribeRequest      = 0x01,
+    kMsgType_ReadRequest           = 0x02,
+    kMsgType_ReportData            = 0x03,
+    kMsgType_WriteRequest          = 0x04,
+    kMsgType_WriteResponse         = 0x05,
+    kMsgType_InvokeCommandRequest  = 0x06,
+    kMsgType_InvokeCommandResponse = 0x07,
+};
+
+class ParserBase
+{
+public:
+    /**
+     *  @brief Initialize a TLVReader to point to the beginning of any tagged element in this request
+     *
+     *  @param [in]  aTagToFind Tag to find in the request
+     *  @param [out] apReader   A pointer to TLVReader, which will be initialized at the specified TLV element
+     *                          on success
+     *
+     *  @return #CHIP_NO_ERROR on success
+     */
+    CHIP_ERROR GetReaderOnTag(const uint64_t aTagToFind, chip::TLV::TLVReader * const apReader) const;
+
+    /**
+     *  @brief Get the TLV Reader
+     *
+     *  @param [in] aReader A pointer to a TLVReader
+     *
+     */
+    void GetReader(chip::TLV::TLVReader * const apReader);
+
+protected:
+    chip::TLV::TLVReader mReader;
+
+    ParserBase();
+
+    template <typename T>
+    CHIP_ERROR GetUnsignedInteger(const uint8_t aContextTag, T * const apLValue) const;
+
+    template <typename T>
+    CHIP_ERROR GetSimpleValue(const uint8_t aContextTag, const chip::TLV::TLVType aTLVType, T * const apLValue) const;
+};
+
+class ListParserBase : public ParserBase
+{
+protected:
+    ListParserBase();
+
+public:
+    /**
+     *  @brief Initialize the parser object with TLVReader
+     *
+     *  @param [in] aReader A pointer to a TLVReader, which should be on the element of the array element
+     *
+     *  @return #CHIP_NO_ERROR on success
+     */
+    CHIP_ERROR Init(const chip::TLV::TLVReader & aReader);
+
+    /**
+     *  @brief Initialize the parser object with TLVReader if context tag exists
+     *
+     *  @param [in] aReader A pointer to a TLVReader, which should be on the element of the array element
+     *  @param [in] aContextTagToFind A context tag it tries to find
+     *
+     *  @return #CHIP_NO_ERROR on success
+     */
+    CHIP_ERROR InitIfPresent(const chip::TLV::TLVReader & aReader, const uint8_t aContextTagToFind);
+
+    /**
+     *  @brief Iterate to next element
+     *
+     *  @return #CHIP_NO_ERROR on success
+     */
+    CHIP_ERROR Next();
+};
+
+class BuilderBase
+{
+public:
+    /**
+     *  @brief Initialize the parser object with TLVReader and ContainerType
```

Comment:
this is copy-paste?  initialize the builder?

Reply:
Thanks for catching up. Just fix it.